### PR TITLE
Checking that symfony installer is up-to-date

### PR DESCRIPTION
**Pull request by jzawadzki**
Hi,

regarding our discussion in #220 and #136 I'm sending first draft of checking if installer is up-to-date.
I have based my solution of how and when composer is checking it.
Please comment and tell if this solution is ok and what should I change to make it better!

Thanks,
JZ

### CODE REVIEW

**Comment by Pierstoval** [comment]
Hehe, we proposed the same feature at the same time :laughing: (#221)

File: src/Symfony/Installer/Application.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Application as ConsoleApplication;
+
+/**
+ * This is main Symfony Installer console application class.
+ */
+class Application extends ConsoleApplication
+{
+    const VERSIONS_URL = 'http://get.sensiolabs.org/symfony.version';
```
Comment (stof) [inline src/Symfony/Installer/Application.php]: I suggest using ``get.symfony.com`` instead (this is the Symfony installer after all), as it also works (and even with https)

File: src/Symfony/Installer/Application.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    public function doRun(InputInterface $input, OutputInterface $output)
+    {
+        $commandName = $this->getCommandName($input);
+
+        if ($this->isInPharMode() && $commandName != 'self-update' && $commandName == 'selfupdate') {
```
Comment (stof) [inline src/Symfony/Installer/Application.php]: should be ``!= 'selfupdate'``

**Comment by jzawadzki** [comment]
@Pierstoval yes - minutes away! :D  2 solutions are still better then no solution!
Good we have @javiereguiluz to decide ;)

File: src/Symfony/Installer/Application.php (hunk 1 above)
Comment (jzawadzki) [inline src/Symfony/Installer/Application.php]: will change to `get.symfony.com` but we still don't use https here 
https://github.com/symfony/symfony-installer/blob/master/src/Symfony/Installer/SelfUpdateCommand.php#L125 ?
Is there any reason

Comment (stof) [inline src/Symfony/Installer/Application.php]: mostly because the last time we worked on the SelfUpdateCommand, SSL was not yet available

Comment (jzawadzki) [inline src/Symfony/Installer/Application.php]: updated to https - thanks for claryifing

File: src/Symfony/Installer/Application.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+ */
+class Application extends ConsoleApplication
+{
+    const VERSIONS_URL = 'https://get.symfony.com/symfony.version';
+
+    public function doRun(InputInterface $input, OutputInterface $output)
+    {
+        $commandName = $this->getCommandName($input);
+
+        if ($this->isInPharMode() && $commandName != 'self-update' && $commandName != 'selfupdate') {
```
Comment (ogizanagi) [inline src/Symfony/Installer/Application.php]:
```diff
- if ($this->isInPharMode() && $commandName != 'self-update' && $commandName != 'selfupdate') {
+ if ($this->isInPharMode() && !in_array($commandName, array('self-update', 'selfupdate'), true) {
```

File: src/Symfony/Installer/Application.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        return 'phar://' === substr(__DIR__, 0, 7);
+    }
+
+    private function checkIfInstallerIsUpdated()
+    {
+        $localVersion = $this->getVersion();
+
+        if (false === $remoteVersion = @file_get_contents(self::VERSIONS_URL)) {
+            // as this is simple checking - we don't here care here if versions file is unavailable
```
Comment (ogizanagi) [inline src/Symfony/Installer/Application.php]:
```diff
- // as this is simple checking - we don't here care here if versions file is unavailable
+ // as this is simple checking - we don't care here if versions file is unavailable
```

(extra `here`)

File: src/Symfony/Installer/Application.php (hunk 3 above)
Comment (jzawadzki) [inline src/Symfony/Installer/Application.php]: changed - thx for review

File: src/Symfony/Installer/Application.php (hunk 4 above)
Comment (jzawadzki) [inline src/Symfony/Installer/Application.php]: fixed  - thank you!

File: src/Symfony/Installer/Application.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ */
+class Application extends ConsoleApplication
+{
+    const VERSIONS_URL = 'https://get.symfony.com/symfony.version';
+
+    public function doRun(InputInterface $input, OutputInterface $output)
+    {
+        $commandName = $this->getCommandName($input);
+
+        if ($this->isInPharMode() && !in_array($commandName, array('self-update', 'selfupdate'), true)) {
```
Comment (javiereguiluz) [inline src/Symfony/Installer/Application.php]: Just asking: shouldn't we also avoid to call this command for `list`, `help` and the other common actions? Maybe we could replace the `!in_array($commandName, array(...))` by `in_array($commandName, array('new', 'demo'))` ?

Comment (jzawadzki) [inline src/Symfony/Installer/Application.php]: At first I was :+1: You are right that those are mostly needed on `new` and `demo`, but after some considerations:

What if we change some command names/parameters and user will get old version on `help`  (different than e.g. on symfony.com site) - that could make some problems I think?

Maybe we should still check version there and set notification but not show it as warning?

Comment (ogizanagi) [inline src/Symfony/Installer/Application.php]: IMHO, commands arguments and options are not likely to change very often, and the `help` command will still be valid for the current installed version. If someone does not have a new option, I think he'll figure out his installer is outdated.

So :+1: about Javier suggestion.

Comment (jzawadzki) [inline src/Symfony/Installer/Application.php]: ok - updated!

**Comment by javiereguiluz** [comment]
@jzawadzki thanks for this feature! (and thanks to @Pierstoval too for working on this feature in another pull request!)

**Pull request closed without merging**